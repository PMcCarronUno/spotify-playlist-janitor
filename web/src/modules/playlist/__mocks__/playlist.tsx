import React from 'react';
export const Playlist = ({ id }: { id: string }) => <div>Playlist: {id} Component</div>;
