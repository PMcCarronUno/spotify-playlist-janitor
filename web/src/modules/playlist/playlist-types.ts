import { Playlist } from "shared/types";

export type DetailedPlaylist = Playlist & {
  skipThreshold?: number;
  ignoreInitialSkips: boolean;
  autoCleanupLimit?: number;
};
