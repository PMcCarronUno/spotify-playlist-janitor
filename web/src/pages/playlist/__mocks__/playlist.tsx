import React from 'react';
export const Playlist = () => <div>Playlist</div>;