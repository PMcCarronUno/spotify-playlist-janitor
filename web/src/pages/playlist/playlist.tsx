import { useParams } from "react-router-dom";
import styled from "styled-components";
import { Playlist as PlaylistInitialise } from "modules/playlist";
import { HEADER_HEIGHT, HEADER_PADDING } from "shared/constants";

export const Playlist = () => {
  let { id } = useParams();
  return (
    <Content>
      <PlaylistInitialise {...{ id }} />
    </Content>
  );
};

const Content = styled.div`
  display: flex;
  justify-content: center;
  height: calc(
    100vh - ${HEADER_HEIGHT} - ${HEADER_PADDING} - ${HEADER_PADDING}
  );
`;

Playlist.displayName = "Playlist";
