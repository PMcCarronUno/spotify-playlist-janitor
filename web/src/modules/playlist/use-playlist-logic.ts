import { useCallback, useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";

import { useDataApi } from "api/data-api";
import { useSpotifyApi } from "api/spotify-api";

import { DetailedPlaylist } from "./playlist-types";
import { useModal } from "shared/components/modal";
import { DEFAULT_IMAGE, HOME } from "shared/constants";

type PlaylistProps = {
  id: string;
};

export const usePlaylistLogic = ({ id }: PlaylistProps) => {
  const {
    getDatabasePlaylist,
    updateDatabasePlaylist,
    deleteDatabasePlaylist,
  } = useDataApi();
  const { getSpotifyPlaylist } = useSpotifyApi();
  const navigate = useNavigate();

  const [notFound, setNotFound] = useState(false);
  const [loading, setLoading] = useState(false);
  const [playlist, setPlaylist] = useState<DetailedPlaylist | undefined>(
    undefined
  );

  const [editSaving, setEditSaving] = useState(false);
  const [editError, setEditError] = useState<string | undefined>(undefined);

  const [deleting, setDeleting] = useState(false);
  const [deleteError, setDeleteError] = useState<string | undefined>(undefined);

  const {
    isOpen: editOpen,
    onOpen: onEditOpen,
    onClose: onEditModalClose,
  } = useModal();
  const {
    isOpen: deleteOpen,
    onOpen: onDeleteOpen,
    onClose: onDeleteClose,
  } = useModal();

  const onEditClose = useCallback(() => {
    setEditError(undefined);
    onEditModalClose();
  }, [setEditError, onEditModalClose]);

  const getPlaylistInfo = useCallback(
    async (id: string) => {
      try {
        setLoading(true);
        setPlaylist(undefined);
        const databasePlaylistsResponse = await getDatabasePlaylist(id);

        const spotifyPlaylistsResponse = await getSpotifyPlaylist(id);

        setPlaylist({
          id: databasePlaylistsResponse.data.id,
          skipThreshold: databasePlaylistsResponse.data.skipThreshold,
          ignoreInitialSkips: databasePlaylistsResponse.data.ignoreInitialSkips,
          autoCleanupLimit: databasePlaylistsResponse.data.autoCleanupLimit,
          name: spotifyPlaylistsResponse.data.name,
          href: spotifyPlaylistsResponse.data.href,
          image: spotifyPlaylistsResponse.data.images[0] ?? DEFAULT_IMAGE,
        });

        setLoading(false);
      } catch (e: any) {
        console.log("error", e)
        if (e?.response?.status === 404) {
          setNotFound(true);
        }
        setLoading(false);
      }
    },
    [setLoading, getDatabasePlaylist, getSpotifyPlaylist, setNotFound]
  );

  useEffect(() => {
    if (!!id) {
      getPlaylistInfo(id);
    }
  }, [getPlaylistInfo, id]);

  const onEditSubmit = useCallback(
    async (event: React.FormEvent<HTMLFormElement>) => {
      event.preventDefault();
      setEditError(undefined);
      const skipThreshold = (event.currentTarget[0] as HTMLInputElement).value;
      const ignoreInitialSkips = (event.currentTarget[1] as HTMLInputElement)
        .checked;
      const autoDeleteTracksAfter = (event.currentTarget[2] as HTMLInputElement)
        .value;

      if (!!autoDeleteTracksAfter && parseInt(autoDeleteTracksAfter) === 0) {
        setEditError("Auto-delete must be empty or greater than 0");
        return;
      }

      try {
        setEditSaving(true);

        const request = {
          skipThreshold: !!skipThreshold ? parseInt(skipThreshold) : undefined,
          ignoreInitialSkips: ignoreInitialSkips,
          autoCleanupLimit: !!autoDeleteTracksAfter
            ? parseInt(autoDeleteTracksAfter)
            : undefined,
        };

        await updateDatabasePlaylist(id, request);
        setEditSaving(false);
        onEditClose();
        await getPlaylistInfo(id);
      } catch (e: any) {
        setEditError(e.response?.data?.message || "Unknown error");
        setEditSaving(false);
      }
    },
    [id, updateDatabasePlaylist, getPlaylistInfo, onEditClose]
  );

  const onDeleteSubmit = useCallback(
    async (event: React.FormEvent<HTMLFormElement>) => {
      event.preventDefault();

      try {
        setDeleting(true);
        await deleteDatabasePlaylist(id);
        setDeleting(false);
        onDeleteClose();
        navigate(HOME)
      } catch (e: any) {
        setDeleteError(e.response?.data?.message || "Unknown error");
        setDeleting(false);
      }
    },
    [id, deleteDatabasePlaylist, navigate, onDeleteClose]
  );

  return {
    loading,
    notFound,
    playlist,
    editOpen,
    onEditOpen,
    onEditClose,
    onEditSubmit,
    editError,
    editSaving,
    deleting,
    deleteOpen, 
    onDeleteOpen, 
    onDeleteClose,
    onDeleteSubmit,
    deleteError,
  } as const;
};
