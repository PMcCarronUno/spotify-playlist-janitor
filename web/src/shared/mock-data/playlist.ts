import { DetailedPlaylist } from "modules/playlist/playlist-types";
import {
  SkippedTrackHistory,
  SkippedTrackTotal,
  Track,
} from "modules/playlist/tabs/playlist-tabs-types";
import { DEFAULT_ALBUM_IMAGE } from "shared/constants";
import {
  databasePlaylists,
  spotifyPlaylists,
  skippedTracks,
  spotifyTracks as mockSpotifyTracks,
} from "shared/mock-data/api";
import { Image } from "shared/types";

const image: Image = {
  height:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.images[0]
      ?.height || 100,
  width:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.images[0]
      ?.width || 100,
  url:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.images[0]
      ?.url || "href",
};

export const playlist: DetailedPlaylist = {
  id: databasePlaylists[0].id,
  href:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.href || "",
  name:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.name || "",
  image: image,
  ignoreInitialSkips: databasePlaylists[0].ignoreInitialSkips,
  autoCleanupLimit: databasePlaylists[0].autoCleanupLimit,
  skipThreshold: databasePlaylists[0].skipThreshold,
};

export const skippedTrackTotals: SkippedTrackTotal[] = [
  {
    id: skippedTracks[0].trackId,
    name: skippedTracks[0].name,
    duration: skippedTracks[0].duration,
    skippedTotal: 3,
    album: skippedTracks[0].album,
    artists: skippedTracks[0].artists,
    image: skippedTracks[0].album.images.reduce((prev, curr) => {
      return prev.height < curr.height ? prev : curr;
    }),
  },
  {
    id: skippedTracks[3].trackId,
    name: skippedTracks[3].name,
    duration: skippedTracks[3].duration,
    skippedTotal: 2,
    album: skippedTracks[3].album,
    artists: skippedTracks[3].artists,
    image: skippedTracks[3].album.images.reduce((prev, curr) => {
      return prev.height < curr.height ? prev : curr;
    }),
  },
];

export const skippedTrackHistory: SkippedTrackHistory[] = skippedTracks.map(
  (skippedTrack) => ({
    id: skippedTrack.trackId,
    name: skippedTrack.name,
    duration: skippedTrack.duration,
    skippedDate: new Date(skippedTrack.skippedDate),
    album: skippedTrack.album,
    artists: skippedTrack.artists,
    image: skippedTrack.album.images.reduce((prev, curr) => {
      return prev.height < curr.height ? prev : curr;
    }),
  })
);

export const spotifyTracks: Track[] = mockSpotifyTracks.map((spotifytrack) => {
  let image: Image = {
    height: 40,
    width: 40,
    url: DEFAULT_ALBUM_IMAGE,
  };

  if (spotifytrack.album.images.length > 0) {
    image = spotifytrack.album.images.reduce((prev, curr) => {
      return prev.height < curr.height ? prev : curr;
    });
  }

  return {
    id: spotifytrack.id,
    name: spotifytrack.name,
    duration: spotifytrack.duration,
    album: spotifytrack.album,
    artists: spotifytrack.artists,
    image,
  };
});
