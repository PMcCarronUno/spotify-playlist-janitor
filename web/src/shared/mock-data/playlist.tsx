import format from "date-fns/format";
import { DetailedPlaylist } from "modules/playlist/playlist-types";
import {
  SkippedTrackAlbum,
  SkippedTrackArtist,
  SkippedTrackHistory,
  SkippedTrackTotal,
  Track,
} from "modules/playlist/tabs/playlist-tabs-types";
import { PiClock } from "react-icons/pi";
import { VscTrash } from "react-icons/vsc";
import { Skeleton, SkeletonTheme } from "shared/components/skeleton";
import { Column } from "shared/components/table";
import { SecondaryText } from "shared/components/typography";
import {
  DEFAULT_ALBUM_IMAGE,
  SKELETON_HIGHLIGHT,
  SKELETON_PLAYLIST_BASE,
} from "shared/constants";
import {
  databasePlaylists,
  spotifyPlaylists,
  skippedTracks,
  spotifyTracks as mockSpotifyTracks,
} from "shared/mock-data/api";
import { Image } from "shared/types";

const image: Image = {
  height:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.images[0]
      ?.height || 100,
  width:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.images[0]
      ?.width || 100,
  url:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.images[0]
      ?.url || "href",
};

export const playlist: DetailedPlaylist = {
  id: databasePlaylists[0].id,
  href:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.href || "",
  name:
    spotifyPlaylists.find((x) => x.id === databasePlaylists[0].id)?.name || "",
  image: image,
  ignoreInitialSkips: databasePlaylists[0].ignoreInitialSkips,
  autoCleanupLimit: databasePlaylists[0].autoCleanupLimit,
  skipThreshold: databasePlaylists[0].skipThreshold,
};

export const skippedTrackTotals: SkippedTrackTotal[] = [
  {
    id: skippedTracks[0].trackId,
    name: skippedTracks[0].name,
    duration: skippedTracks[0].duration,
    skippedTotal: 3,
    album: skippedTracks[0].album,
    artists: skippedTracks[0].artists,
    image: skippedTracks[0].album.images.reduce((prev, curr) => {
      return prev.height < curr.height ? prev : curr;
    }),
  },
  {
    id: skippedTracks[3].trackId,
    name: skippedTracks[3].name,
    duration: skippedTracks[3].duration,
    skippedTotal: 2,
    album: skippedTracks[3].album,
    artists: skippedTracks[3].artists,
    image: skippedTracks[3].album.images.reduce((prev, curr) => {
      return prev.height < curr.height ? prev : curr;
    }),
  },
];

export const skippedTrackHistory: SkippedTrackHistory[] = skippedTracks.map(
  (skippedTrack) => ({
    id: skippedTrack.trackId,
    name: skippedTrack.name,
    duration: skippedTrack.duration,
    skippedDate: new Date(skippedTrack.skippedDate),
    album: skippedTrack.album,
    artists: skippedTrack.artists,
    image: skippedTrack.album.images.reduce((prev, curr) => {
      return prev.height < curr.height ? prev : curr;
    }),
  })
);

export const spotifyTracks: Track[] = mockSpotifyTracks.map((spotifytrack) => {
  let image: Image = {
    height: 40,
    width: 40,
    url: DEFAULT_ALBUM_IMAGE,
  };

  if (spotifytrack.album.images.length > 0) {
    image = spotifytrack.album.images.reduce((prev, curr) => {
      return prev.height < curr.height ? prev : curr;
    });
  }

  return {
    id: spotifytrack.id,
    name: spotifytrack.name,
    duration: spotifytrack.duration,
    album: spotifytrack.album,
    artists: spotifytrack.artists,
    image,
  };
});

const skeleton = (
  <SkeletonTheme
    baseColor={SKELETON_PLAYLIST_BASE}
    highlightColor={SKELETON_HIGHLIGHT}
    height="100%"
  >
    <Skeleton />
  </SkeletonTheme>
);

export const imageColumn: Column = {
  label: "",
  accessor: "image",
  primary: true,
  render: (image: Image, loading?: boolean) => (
    <div className="centered">
      {loading ? (
        <div style={{ height: 40, width: 40 }}>{skeleton}</div>
      ) : (
        <img alt="" width={40} height={40} src={image.url} />
      )}
    </div>
  ),
};

export const titleColumn: Column = {
  label: "Title",
  accessor: "name",
  sortable: true,
  primary: true,
  render: (value: string, loading?: boolean) => (loading ? skeleton : value),
};

export const artistColumn: Column = {
  label: "Artist",
  accessor: "artists",
  sortable: true,
  render: (artists: SkippedTrackArtist[], loading?: boolean) => (
    <>
      {loading
        ? skeleton
        : artists.map((artist, index) => (
            <>
              {`${index === 0 ? "" : ", "}`}
              {!!artist.href ? (
                <a href={artist.href} target="_blank" rel="noreferrer">
                  {artist.name}
                </a>
              ) : (
                <SecondaryText>{artist.name}</SecondaryText>
              )}
            </>
          ))}
    </>
  ),
};

export const albumColumn: Column = {
  label: "Album",
  accessor: "album",
  sortable: true,
  render: (album: SkippedTrackAlbum, loading?: boolean) =>
    loading ? (
      skeleton
    ) : !!album.href ? (
      <a href={album.href} target="_blank" rel="noreferrer">
        {album.name}
      </a>
    ) : (
      <SecondaryText>{album.name}</SecondaryText>
    ),
};

export const durationColumn: Column = {
  label: <PiClock />,
  accessor: "duration",
  sortable: true,
  render: (millis: number, loading?: boolean) => {
    if (loading) {
      return skeleton;
    }
    var minutes = Math.floor(millis / 60000);
    var seconds = parseInt(((millis % 60000) / 1000).toFixed(0));
    return seconds === 60
      ? minutes + 1 + ":00"
      : minutes + ":" + (seconds < 10 ? "0" : "") + seconds;
  },
};

export const totalSkippedColumn: Column = {
  label: "Date Skipped",
  accessor: "skippedDate",
  sortable: true,
  sortbyOrder: "desc",
  render: (date: Date, loading?: boolean) =>
    loading
      ? skeleton
      : format(date, "MMM dd yyyy") + " at " + format(date, "HH:mm:ss"),
};

export const dateSkippedColumn: Column = {
  label: "Total Skipped",
  accessor: "skippedTotal",
  sortable: true,
  sortbyOrder: "desc",
  render: (value: string, loading?: boolean) => (loading ? skeleton : value),
};

export const getDeleteColumn = (onDeleteClick: (id: string) => void) => ({
  label: "",
  accessor: "id",
  render: (id: string, loading?: boolean) =>
    loading ? (
      <></>
    ) : (
      <div className="centered">
        <VscTrash
          style={{ cursor: "pointer" }}
          onClick={() => onDeleteClick(id)}
        />
      </div>
    ),
});
