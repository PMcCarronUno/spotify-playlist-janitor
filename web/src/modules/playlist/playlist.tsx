import styled from "styled-components";
import {
    SKELETON_PLAYLIST_BASE,
    SKELETON_HIGHLIGHT,
    WHITE,
} from "shared/constants";
import { usePlaylistLogic } from "./use-playlist-logic";
import { Menu, MenuItem } from "shared/components/menu";
import { VscEdit, VscEllipsis, VscLinkExternal, VscTrash } from "react-icons/vsc";
import { SecondaryText, Text, Title } from "shared/components/typography";
import { Skeleton, SkeletonTheme } from "shared/components/skeleton";
import { Modal } from "shared/components/modal";
import { EditPlaylistModalView } from "./modal";
import { DeletePlaylistModalView } from "./modal/delete-playlist-modal";
import { PlaylistTabs } from "./tabs/playlist-tabs";


type PlaylistProps = {
    id: string;
};

const skeleton = <SkeletonTheme
    baseColor={SKELETON_PLAYLIST_BASE}
    highlightColor={SKELETON_HIGHLIGHT}
    height="100%"
>
    <Skeleton />
</SkeletonTheme>;

export const Playlist = ({ id }: PlaylistProps) => {
    const {
        loading,
        notFound,
        playlist,
        editOpen,
        onEditOpen,
        onEditClose,
        onEditSubmit,
        editError,
        editSaving,
        deleting,
        deleteOpen,
        onDeleteOpen,
        onDeleteClose,
        onDeleteSubmit,
        deleteError,
    } = usePlaylistLogic({ id });

    if (notFound) {
        return <PageContainer>
            <Text {...{
                id: "not-found",
                "data-testid": "not-found",
            }}>
                Not Found
            </Text>
        </PageContainer>;
    }

    return (
        <PageContainer>
            <div className="header">
                <div className="playlist">
                    <div className="playlist-image" id="playlist-image">
                        {loading ?
                            skeleton
                            : (
                                <img
                                    width={240}
                                    height={240}
                                    src={playlist?.image?.url}
                                    alt={playlist?.name}
                                />
                            )}
                    </div>
                    <div className="playlist-details">
                        <Title className={`playlist-title${loading ? "-loading" : ""}`}>
                            {!!playlist
                                ?
                                <>
                                    {playlist?.name}
                                    <a className="icon-link"
                                        href={playlist?.href}
                                        target="_blank"
                                        rel="noreferrer">
                                        <VscLinkExternal style={{ cursor: "pointer", fontSize: 28 }} />
                                    </a>
                                </>
                                :
                                skeleton
                            }
                        </Title>
                        <div
                            className={`playlist-details-item${loading ? "-loading" : ""}`}
                        >
                            {playlist?.skipThreshold ? (
                                <>
                                    <SecondaryText>Skip Threshold:&nbsp;</SecondaryText>
                                    <Text>{playlist?.skipThreshold} s</Text>
                                </>
                            ) : skeleton}
                        </div>

                        <div
                            className={`playlist-details-item${loading ? "-loading" : ""}`}
                        >
                            {playlist?.skipThreshold ? (
                                <>
                                    <SecondaryText>Ignore initial skips:&nbsp;</SecondaryText>
                                    <Text>{playlist?.ignoreInitialSkips ? "True" : "False"}</Text>
                                </>
                            ) : skeleton}
                        </div>

                        <div
                            className={`playlist-details-item${loading ? "-loading" : ""}`}
                        >
                            {playlist ? (
                                <>
                                    <SecondaryText>Auto-delete after:&nbsp;</SecondaryText>
                                    {playlist?.autoCleanupLimit ? (
                                        <>
                                            <Text>{playlist?.autoCleanupLimit}&nbsp;</Text>
                                            <SecondaryText>skips</SecondaryText>
                                        </>
                                    ) : (
                                        <Text>No auto-delete</Text>
                                    )}
                                </>
                            ) : skeleton}
                        </div>
                    </div>
                </div>
                <div className="playlist-menu">
                    <Menu
                        menuButton={
                            <div>
                                <VscEllipsis style={{ cursor: "pointer", fontSize: 28 }} />
                            </div>
                        }
                    >
                        <MenuItem {...{ onClick: onEditOpen }}>
                            <VscEdit style={{ cursor: "pointer", marginRight: 8 }} />
                            Edit
                        </MenuItem>
                        <MenuItem {...{ onClick: onDeleteOpen }}>
                            <VscTrash style={{ cursor: "pointer", marginRight: 8 }} />
                            Delete
                        </MenuItem>
                    </Menu>
                </div>
            </div>

            <div className="content">
                <PlaylistTabs {...{ id, loading }} />
            </div>

            <Modal
                {...{
                    isOpen: editOpen,
                    onClose: onEditClose,
                    label: "Edit Playlist Modal",
                }}
            >
                <EditPlaylistModalView
                    {...{
                        onSubmit: onEditSubmit,
                        playlist,
                        modalSaving: editSaving,
                        modalError: editError,
                        onModalClose: onEditClose,
                    }}
                />
            </Modal>
            <Modal
                {...{
                    isOpen: deleteOpen,
                    onClose: onDeleteClose,
                    label: "Delete Playlist Modal",
                }}
            >
                <DeletePlaylistModalView
                    {...{
                        onSubmit: onDeleteSubmit,
                        playlist,
                        modalSaving: deleting,
                        modalError: deleteError,
                        onModalClose: onDeleteClose,
                    }}
                />
            </Modal>
        </PageContainer>
    );
};

const PageContainer = styled.div`
  display: flex;
  flex-direction: column;
  color: ${WHITE};
  align-items: center;

  width: 100%;
  max-width: 1825px;
  padding-top: 16px;

  .header {
    display: flex;
    flex-direction: row;
    justify-content: space-around;
    width: 100%;

    margin-bottom: 24px;
  }

  .playlist {
    display: flex;
    flex-direction: row;
  }

  .playlist-image {
    width: 240px;
    height: 240px;
  }

  .playlist-details {
    padding-left: 32px;

    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    align-items: flex-start;

    width: 400px;
  }

  .playlist-title {
    margin-bottom: 24px;
  }

  .playlist-title-loading {
    height: 42px;
    width: 360px;
    margin-bottom: 24px;
  }

  .playlist-details-item {
    margin-bottom: 12px;

    display: flex;
  }

  .playlist-details-item-loading {
    height: 22px;
    width: 180px;
    margin-bottom: 12px;
  }

  .content {
    width: 100%;
  }

  .icon-link{
    color: ${WHITE};
    margin-left: 12px;

    svg {
        margin-bottom: -3px;
    }
  }
`;

Playlist.displayName = "Playlist";
