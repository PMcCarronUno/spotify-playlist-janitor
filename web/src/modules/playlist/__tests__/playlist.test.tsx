import {
  fireEvent,
  getByTestId,
  queryByTestId,
  render,
  waitFor,
} from "@testing-library/react";

import { Playlist } from "../playlist";
import { usePlaylistLogic } from "../use-playlist-logic";
import {
  playlist
} from "shared/mock-data/playlist";

jest.mock("../use-playlist-logic");
jest.mock("../modal");
jest.mock("../tabs");

describe("<Playlist />", () => {
  let container: HTMLElement;
  let mockUsePlaylistLogic: any;

  beforeEach(async () => {

    mockUsePlaylistLogic = {
      playlist,
      loading: false,
      notFound: false,
      editOpen: false,
      onEditOpen: jest.fn(),
      onEditClose: jest.fn(),
      onEditSubmit: jest.fn(),
      editError: "",
      editSaving: false,
      showSpotifyAuthModal: false,
    };

    jest.mocked(usePlaylistLogic).mockImplementation(() => mockUsePlaylistLogic);

    await waitFor(() => {
      ({ container } = render(<Playlist {...{id: playlist.id}}/>));
    });
  });

  it("should render Playlist component", () => {
    expect(container.firstChild).toMatchSnapshot();

    expect(container.getElementsByClassName("react-loading-skeleton").length).toBe(0)
  });

  it("should render Playlist component in loading state", async () => {
    mockUsePlaylistLogic = {
      ...mockUsePlaylistLogic,
      playlist: undefined,
      loading: true,
    };
    jest.mocked(usePlaylistLogic).mockImplementation(() => mockUsePlaylistLogic);

    await waitFor(() => {
      ({ container } = render(<Playlist {...{id: playlist.id}}/>));
    });

    expect(container.firstChild).toMatchSnapshot();

    expect(container.getElementsByClassName("react-loading-skeleton").length).toBe(5)
  });

  it("should render Home component with Not Found", async () => {
    mockUsePlaylistLogic = {
      ...mockUsePlaylistLogic,
      notFound: true,
    };
    jest.mocked(usePlaylistLogic).mockImplementation(() => mockUsePlaylistLogic);

    await waitFor(() => {
      ({ container } = render(<Playlist {...{id: playlist.id}}/>));
    });

    expect(container.firstChild).toMatchSnapshot();

    expect(getByTestId(container, "not-found")).toBeTruthy();
  });
});
