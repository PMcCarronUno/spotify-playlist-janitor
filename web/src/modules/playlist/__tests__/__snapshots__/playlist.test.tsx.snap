// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Playlist /> should render Home component with Not Found 1`] = `
<div
  class="sc-gLLuHO efEqvT"
>
  <div
    class="sc-hLQTFJ joXFND"
    data-testid="not-found"
    id="not-found"
  >
    Not Found
  </div>
</div>
`;

exports[`<Playlist /> should render Playlist component 1`] = `
<div
  class="sc-gLLuHO efEqvT"
>
  <div
    class="header"
  >
    <div
      class="playlist"
    >
      <div
        class="playlist-image"
      >
        <img
          alt="Playlist 1"
          height="240"
          src="https://mosaic.scdn.co/101"
          width="240"
        />
      </div>
      <div
        class="playlist-details"
      >
        <div
          class="sc-guDMob cuZJcO playlist-title"
        >
          Playlist 1
        </div>
        <div
          class="playlist-details-item"
        >
          <div
            class="sc-eDLKEg jzLtSt"
          >
            Skip Threshold: 
          </div>
          <div
            class="sc-hLQTFJ joXFND"
          >
            10
             s
          </div>
        </div>
        <div
          class="playlist-details-item"
        >
          <div
            class="sc-eDLKEg jzLtSt"
          >
            Ignore initial skips: 
          </div>
          <div
            class="sc-hLQTFJ joXFND"
          >
            False
          </div>
        </div>
        <div
          class="playlist-details-item"
        >
          <div
            class="sc-eDLKEg jzLtSt"
          >
            Auto-delete after: 
          </div>
          <div
            class="sc-hLQTFJ joXFND"
          >
            10
             
          </div>
          <div
            class="sc-eDLKEg jzLtSt"
          >
            skips
          </div>
        </div>
      </div>
    </div>
    <div
      class="playlist-menu"
    >
      <div>
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          style="cursor: pointer; font-size: 28px;"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 8a1 1 0 1 1-2 0 1 1 0 0 1 2 0zm5 0a1 1 0 1 1-2 0 1 1 0 0 1 2 0zm5 0a1 1 0 1 1-2 0 1 1 0 0 1 2 0z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Playlist /> should render Playlist component in loading state 1`] = `
<div
  class="sc-gLLuHO efEqvT"
>
  <div
    class="header"
  >
    <div
      class="playlist"
    >
      <div
        class="playlist-image"
      >
        <span
          aria-busy="true"
          aria-live="polite"
        >
          <span
            class="react-loading-skeleton"
            style="height: 100%; --base-color: #111111; --highlight-color: #444;"
          >
            ‌
          </span>
          <br />
        </span>
      </div>
      <div
        class="playlist-details"
      >
        <div
          class="sc-guDMob cuZJcO playlist-title-loading"
        >
          <span
            aria-busy="true"
            aria-live="polite"
          >
            <span
              class="react-loading-skeleton"
              style="height: 100%; --base-color: #111111; --highlight-color: #444;"
            >
              ‌
            </span>
            <br />
          </span>
        </div>
        <div
          class="playlist-details-item-loading"
        >
          <span
            aria-busy="true"
            aria-live="polite"
          >
            <span
              class="react-loading-skeleton"
              style="height: 100%; --base-color: #111111; --highlight-color: #444;"
            >
              ‌
            </span>
            <br />
          </span>
        </div>
        <div
          class="playlist-details-item-loading"
        >
          <span
            aria-busy="true"
            aria-live="polite"
          >
            <span
              class="react-loading-skeleton"
              style="height: 100%; --base-color: #111111; --highlight-color: #444;"
            >
              ‌
            </span>
            <br />
          </span>
        </div>
        <div
          class="playlist-details-item-loading"
        >
          <span
            aria-busy="true"
            aria-live="polite"
          >
            <span
              class="react-loading-skeleton"
              style="height: 100%; --base-color: #111111; --highlight-color: #444;"
            >
              ‌
            </span>
            <br />
          </span>
        </div>
      </div>
    </div>
    <div
      class="playlist-menu"
    >
      <div>
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          style="cursor: pointer; font-size: 28px;"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 8a1 1 0 1 1-2 0 1 1 0 0 1 2 0zm5 0a1 1 0 1 1-2 0 1 1 0 0 1 2 0zm5 0a1 1 0 1 1-2 0 1 1 0 0 1 2 0z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
